File: Framework/Php/framework/framework/net/smartsocket.php
<?php

/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */

namespace Smartkernel\Framework\Net {

    use Smartkernel\Framework\Text\SmartEncoding;

/*
     * 套接字
     */

    class SmartSocket {
        /*
         * 发送信息
         * @param string $ip，端点IP
         * @param string $port，端点Port
         * @param string $requestText，消息
         * @param string $flgReceive，是否接受消息
         * @param string $encoding，数据编码方式
         * @return string,发送消息
         * */

        public static function send($ip, $port, $requestText, $flgReceive = true, $encoding = null) {
            $socket = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
            $connection = socket_connect($socket, $ip, $port);

            $encoding = $encoding == null ? SmartEncoding::DefaultEncoding : $encoding;

            socket_write($socket, iconv($encoding, SmartEncoding::DefaultEncoding, $requestText));
            socket_shutdown($socket, STREAM_SHUT_WR);
            if ($flgReceive === false) {
                return "";
            }
            $responseText = '';
            do {
                $buffer = socket_read($socket, 1024);
                $response .= $buffer;
            } while (!empty($buffer));
            socket_shutdown($socket, STREAM_SHUT_RD);
            return iconv(SmartEncoding::DefaultEncoding, $encoding, $responseText);
        }

        /*
         * 监听消息
         * @param string $ip，端点IP
         * @param string $port，端点Port
         * @param string $interact，交互处理
         * @param string $encoding，数据编码方式
         * */

        function listen($ip, $port, $interact, $encoding = null) {
            $serverSocket = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
            socket_bind($serverSocket, $ip, $port);
            socket_listen($serverSocket);

            $encoding = $encoding == null ? SmartEncoding::DefaultEncoding : $encoding;
            while (true) {
                $socket = socket_accept($serverSocket);
                if ($socket) {
                    $requestText = '';
                    do {
                        $buffer = socket_read($socket, 1024);
                        $requestText .= $buffer;
                    } while (!empty($buffer));
                    socket_shutdown($socket, STREAM_SHUT_RD);
                    $requestText = iconv(SmartEncoding::DefaultEncoding, $encoding, $requestText);
                    $responseText = $interact($requestText);
                    socket_write($socket, iconv($encoding, SmartEncoding::DefaultEncoding, $responseText));
                    socket_shutdown($socket, STREAM_SHUT_WR);
                }
            }
        }

    }

}

File: Framework/CSharp/Framework/Framework/Net/SmartSocket.cs
/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */
using Smartkernel.Framework.Text;
using System;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;

namespace Smartkernel.Framework.Net
{
    /// <summary>
    /// 套接字
    /// </summary>
    public static class SmartSocket
    {
        /// <summary>
        /// 获得远程IP
        /// </summary>
        /// <param name="socket">套接字</param>
        /// <returns>结果</returns>
        public static string GetRemoteEndPointIP(Socket socket)
        {
            SocketAddress socketAddress = socket.RemoteEndPoint.Serialize();

            return socketAddress[4].ToString() + "." + socketAddress[5].ToString() + "." + socketAddress[6].ToString() + "." + socketAddress[7].ToString();
        }

        /// <summary>
        /// 获得远程端口
        /// </summary>
        /// <param name="socket">套接字</param>
        /// <returns>结果</returns>
        public static int GetRemoteEndPointPort(Socket socket)
        {
            SocketAddress socketAddress = socket.RemoteEndPoint.Serialize();

            return socketAddress[2] * 256 + socketAddress[3];
        }

        /// <summary>
        /// 发送信息
        /// </summary>
        /// <param name="ipEndPoint">端点</param>
        /// <param name="requestText">消息</param>
        /// <param name="flgReceive">是否接受消息</param>
        /// <param name="encoding">数据编码方式</param>
        /// <returns>结果</returns>
        public static string Send(IPEndPoint ipEndPoint, string requestText, bool flgReceive = true, Encoding encoding = null)
        {
            encoding = encoding ?? SmartEncoding.Default;

            var responseText = "";

            using (Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
            {
                socket.Connect(ipEndPoint);

                socket.Send(encoding.GetBytes(requestText));
                socket.Shutdown(SocketShutdown.Send);

                if (flgReceive)
                {
                    using (MemoryStream bufferedReader = new MemoryStream())
                    {
                        int length = -1;
                        byte[] bytes = new byte[1024];
                        while ((length = socket.Receive(bytes)) > 0)
                        {
                            bufferedReader.Write(bytes, 0, length);
                        }
                        var array = bufferedReader.ToArray();
                        responseText = encoding.GetString(array, 0, array.Length);
                        socket.Shutdown(SocketShutdown.Receive);
                    }
                }
            }

            return responseText;
        }

        /// <summary>
        /// 监听信息
        /// </summary>
        /// <param name="ipEndPoint">端点</param>
        /// <param name="interact">交互处理</param>
        /// <param name="listenQueueLimit">监听队列限制</param>
        /// <param name="encoding">数据编码方式</param>
        /// <returns>结果</returns>
        public static void Listen(IPEndPoint ipEndPoint, Func<string, string> interact, int listenQueueLimit = 100, Encoding encoding = null)
        {
            encoding = encoding ?? SmartEncoding.Default;

            using (Socket serverSocket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
            {
                serverSocket.Bind(ipEndPoint);
                serverSocket.Listen(listenQueueLimit);
                while (true)
                {
                    Socket socket = serverSocket.Accept();

                    new Task(delegate
                    {
                        using (MemoryStream bufferedReader = new MemoryStream())
                        {
                            int length = -1;
                            byte[] bytes = new byte[1024];
                            while ((length = socket.Receive(bytes)) > 0)
                            {
                                bufferedReader.Write(bytes, 0, length);
                            }
                            var array = bufferedReader.ToArray();
                            var requestText = encoding.GetString(array, 0, array.Length);
                            socket.Shutdown(SocketShutdown.Receive);
                            var responseText = interact(requestText);
                            if (!string.IsNullOrWhiteSpace(responseText))
                            {
                                socket.Send(encoding.GetBytes(responseText));
                                socket.Shutdown(SocketShutdown.Send);
                            }
                        }
                    }).Start();
                }
            }
        }
    }
}
